src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2133,32 +2133,103 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    unsigned overwrittenOpNum = 0;
+    LIR::Use use;
+    if (LIR::AsRange(compiler->compCurBB).TryGetUse(node, &use))
     {
-        // 132 form: op1 = (op1 * op3) + [op2]
+        overwrittenOpNum = node->GetOverwrittenOpNumForFMA(use.User(), op1, op2, op3);
+    }
 
-        ins    = (instruction)(ins - 1);
-        op1Reg = op1->GetRegNum();
-        op2Reg = op3->GetRegNum();
-        op3    = op2;
+    // Intrinsics with CopyUpperBits semantics must have op1 as target
+    if (overwrittenOpNum == 1 || copiesUpperBits)
+    {
+        if (op2->isContained() || op2->IsRegOptional())
+        {
+            // op1 = (op1 * [op2]) + op3
+            // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+            ins    = (instruction)(ins - 1);
+            op1Reg = op1->GetRegNum();
+            op2Reg = op3->GetRegNum();
+            op3    = op2;
+        }
+        else
+        {
+            // op1 = (op1 * op2) + [op3]
+            // 213 form: XMM1 = (XMM2 * XMM1) + [XMM3]
+            op1Reg = op1->GetRegNum();
+            op2Reg = op2->GetRegNum();
+        }
+        isCommutative = copiesUpperBits;
     }
-    else if (op1->isContained() || op1->isUsedFromSpillTemp())
+    else if (overwrittenOpNum == 3)
     {
-        // 231 form: op3 = (op2 * op3) + [op1]
-
+        // 231 form: XMM1 = (XMM2 * [XMM3]) + XMM1
         ins    = (instruction)(ins + 1);
         op1Reg = op3->GetRegNum();
-        op2Reg = op2->GetRegNum();
-        op3    = op1;
+        if (op1->isContained())
+        {
+            // op3 = ([op1] * op2) + op3
+            op2Reg = op2->GetRegNum();
+            op3    = op1;
+        }
+        else
+        {
+            // op3 = (op1 * [op2]) + op3
+            op2Reg = op1->GetRegNum();
+            op3    = op2;
+        }
+    }
+    else if (overwrittenOpNum == 2)
+    {
+        if (op1->isContained())
+        {
+            // op2 = ([op1] * op2) + op3
+            // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+            ins    = (instruction)(ins - 1);
+            op1Reg = op2->GetRegNum();
+            op2Reg = op3->GetRegNum();
+            op3    = op1;
+        }
+        else
+        {
+            // op2 = (op1 * op2) + [op3]
+            // 213 form: XMM1 = (XMM2 * XMM1) + [XMM3]
+            op1Reg = op2->GetRegNum();
+            op2Reg = op1->GetRegNum();
+        }
     }
     else
     {
-        // 213 form: op1 = (op2 * op1) + [op3]
-
-        op1Reg = op1->GetRegNum();
-        op2Reg = op2->GetRegNum();
-
-        isCommutative = !copiesUpperBits;
+        assert(overwrittenOpNum == 0);
+        if (op1->isContained())
+        {
+            // op2 = ([op1] * op2) + op3
+            // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+            ins    = (instruction)(ins - 1);
```

Comment:
Since you do `(instruction)(ins +/- 1)` often, I would recommend to just have something like `instruction _132form = (ins - 1)` etc. at the top and use it here and elsewhere.